src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -349,14 +349,32 @@ fn set_memory_resources(cg: &cgroups::Cgroup, memory: &LinuxMemory, update: bool
         mem_controller.set_kmem_limit(-1)?;
     }
 
-    set_resource!(mem_controller, set_limit, memory, limit);
-    set_resource!(mem_controller, set_soft_limit, memory, reservation);
-    set_resource!(mem_controller, set_kmem_limit, memory, kernel);
-    set_resource!(mem_controller, set_tcp_limit, memory, kernel_tcp);
+    // If the memory update is set to -1 we should also
+    // set swap to -1, it means unlimited memory.
+    let mut swap = memory.swap.unwrap_or(0);
+    if memory.limit == Some(-1) {
+        swap = -1;
+    }
 
-    if let Some(swap) = memory.swap {
-        // set memory swap
-        let swap = if cg.v2() {
+    if memory.limit.is_some() && swap != 0 {
+        let memstat = get_memory_stats(cg)
+            .into_option()
+            .ok_or_else(|| anyhow!("failed to get the cgroup memory stats"))?;
+        let memusage = memstat.get_usage();
+
+        // When update memory limit, we should adapt the write sequence
+        // for memory and swap memory, so it won't fail because the new
+        // value and the old value don't fit kernel's validation.
```

Comment:
Which validation does it fail at? It is a bit confusing having to set two values in two different orders. When we set a swap limit, it must be larger than or equal to the current memory limit?

Reply:
It's  not only depend on the new setting memory limit, but it also depended on the current set memory limit, suppose the current memory limit is large than the new swap, than you cannot set the swap first, instead we should set the memory limit first and then set the swap, otherwise the kernel would complain and refused to set; on the other hand, if the current memory limit is larger than the new swap, then we should set the memory limit first and then set the swap.

Reply:
@bergwolf Add the detail comments.